### PR TITLE
Added library section to the sidebar

### LINKED ISSUE #692: Improving the sidebar
**Original sidebar concepts:**

sidebar concept 84311910-1bb85080-ab32-11ea-933c-2d76c7b

### PR DESCRIPTION
**Pull request by oleitao**
In this PR i added on Settings->Preferences an option to enable to add on sidebar a new section named QuickAccess.  Every files will be Pin into that section and UnPinned from there.

Closes #692.

### CODE REVIEW

**Comment by XPoppyX** [comment]
I think other changes should be made so that the quick access section can be useful, now it's just another "favorites" but without "Home" and the recycle bin.

**Proposal?**
Maybe "Favorites" could be renamed to "Library" or something like that, to show only documents, downloads and others. And in "Quick access" show only other pinned folders

**Comment by oleitao** [comment]
> I think other changes should be made so that the quick access section can be useful, now it's just another "favorites" but without "Home" and the recycle bin.
> 
> **Proposal?**
> Maybe "Favorites" could be renamed to "Library" or something like that, to show only documents, downloads and others. And in "Quick access" show only other pinned folders

On my last PR i developed "Favorites" and collapsible section, based on reviewers suggestions. After some conversation we agree as future work to develop Quick Access section with pin/unpin folders. I suggest you to give your proposal to reviewers in order to validate that.

**Comment by yaichenbaum** [comment]
@oleitao As discussed in #692, we don't have plans for a Quick Access section, the plan is to have a dedicated page instead for recent folders and files. The goal of the sidebar is to provide links to the most common locations that users visit, if something is constantly visited, it's assumed that the user will pin it to the favorites section.

**Comment by oleitao** [comment]
> @oleitao As discussed in #692, we don't have plans for a Quick Access section, the plan is to have a dedicated page instead for recent folders and files. The goal of the sidebar is to provide links to the most common locations that users visit, if something is constantly visited, it's assumed that the user will pin it to the favorites section.

Sorry, for my of feature misunderstanding  :sweat_smile:

**Comment by yaichenbaum** [comment]
@oleitao Sorry for not explaining it clearly, the work in this PR does look amazing 😂. We do plan on adding an optional section for library locations, would you like to work on this? This section would be turned off by default, but there will be a setting to turn it on.

**Comment by oleitao** [comment]
> @oleitao Sorry for not explaining it clearly, the work in this PR does look amazing 😂. We do plan on adding an optional section for library locations, would you like to work on this? This section would be turned off by default, but there will be a setting to turn it on.

yes @yair100 i can take it. but @yair100 can you explain me better the requirements to implement?

- [ ] adding an optional section for library locations (just that?)

Yesterday, before you tell me i don't need to implement QuickAccess section i already had that developed. So, you can check if that can be useful or if not discard (that changes were based on @XPoppyX porposal)

**Comment by yaichenbaum** [comment]
@oleitao The goal is to have an optional library section, this section will be right after the favorites section when turned on. We do have plans for a recent items page so we might be able to reuse some of the quick access code for this page.

**Comment by yaichenbaum** [comment]
@oleitao It doesn't look like the default items are being added to the sidebar anymore, besides for the library section, the goal is to pin a couple of common locations to the favorites section by defualt.

**Comment by yaichenbaum** [comment]
@oleitao The default items are working now, but I don't see anything in the library section.

**Comment by oleitao** [comment]
> 
> 
> @oleitao The default items are working now, but I don't see anything in the library section.

@yair100 i have two questions:

- [ ] what should be placed on library section?
- [ ] Do you have any website or tool to get UWP icons? i try Univsersal Character Map but my icons are not well loaded.

**Comment by yaichenbaum** [comment]
> what should be placed on library section?

@oleitao Do you mean which libraries, or did you mean which icon?

**Comment by oleitao** [comment]
> > what should be placed on library section?
> 
> @oleitao Do you mean which libraries, or did you mean which icon?

yes. i mean what should be the content of that section? And if you can tell me where i can get an icon to use there?

**Comment by yaichenbaum** [comment]
@oleitao For now, I think documents, downloads, music, pictures and videos should be enough. Later on, we can detect all libraries.

**Comment by yaichenbaum** [comment]
@oleitao Can you solve the merge conflicts?

**Comment by oleitao** [comment]
> @oleitao Can you solve the merge conflicts?

@yair100 i could but i dont have write access

File: Files/DataModels/SidebarPinnedModel.cs.bak [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+            }
+            else
+            {
+                iconCode = "\uea55";
+            }
+
+            return iconCode;
+        }
+    }
+}
```
Comment (yaichenbaum) [inline Files/DataModels/SidebarPinnedModel.cs.bak:430]: Is this file necessary?

Comment (oleitao) [reply Files/DataModels/SidebarPinnedModel.cs.bak:430]: no, sorry i used WinMerge and forget to remove that file

File: Files/ViewModels/SettingsViewModel.cs [hunk 2]
```diff
@@ -438,6 +438,18 @@ public bool ShowConfirmDeleteDialog
             set => Set(value);
         }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether [show library switch].
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if [show library section]; otherwise, <c>false</c>.
+        /// </value>
```
Comment (yaichenbaum) [inline Files/ViewModels/SettingsViewModel.cs:446]:
```suggestion
        /// <summary>
        /// Gets or sets a value indicating whether or not to show the library section on the sidebar.
        /// </summary>

**Comment by yaichenbaum** [comment]
@oleitao The last change is that the desktop, documents and downloads folders should be pinned to the favorites section the first time the user runs the app.

File: Files/Filesystem/ListedItem.cs [hunk 3]
```diff
@@ -217,9 +217,7 @@ public ListedItem(string folderRelativeId, string dateReturnFormat = null)
         }
 
         public ListedItem()
-        {
-
-        }
+        { }
```
Comment (yaichenbaum) [inline Files/Filesystem/ListedItem.cs:220]:
```suggestion
        { 
        
        }
```

File: Files/Filesystem/ListedItem.cs [hunk 4]
```diff
@@ -292,7 +290,7 @@ public override string ToString()
         public bool IsShortcutItem => this is ShortcutItem;
         public bool IsLinkItem => IsShortcutItem && ((ShortcutItem)this).IsUrl;
 
-        public bool IsPinned => App.SidebarPinnedController.Model.Items.Contains(itemPath);
+        public bool IsPinned => App.SidebarPinnedController.Model.LibraryItems.Contains(itemPath);
```
Comment (yaichenbaum) [inline Files/Filesystem/ListedItem.cs:293]: What's the goal of this change?

File: Files/Filesystem/ListedItem.cs (hunk 3 above)
Comment (oleitao) [reply Files/Filesystem/ListedItem.cs]: That was to avoid error on GitHub repository code checker

**Comment by yaichenbaum** [comment]
@oleitao The library section seems to be empty now.

File: Files/Filesystem/ListedItem.cs (hunk 4 above)
Comment (oleitao) [reply Files/Filesystem/ListedItem.cs:295]: Yes, i forget to change LibraryItems to FavoriteItems

**Comment by oleitao** [comment]
> @oleitao The library section seems to be empty now.

I don't know whats happen but here library section is loaded alright. Delete App previous configuration (C:\Users\XXXXXX\AppData\Local\Packages\FilesDev_et10x9a9vyk8t\) and try again.

File: Files/DataModels/SidebarPinnedModel.cs.bak (hunk 1 above)
Comment (yaichenbaum) [reply Files/DataModels/SidebarPinnedModel.cs.bak:430]: @oleitao No problem, just make sure to remove it before we merge the PR.

**Review by yair100 — DISMISSED** [review]
Great work! LGTM

**Comment by yaichenbaum** [comment]
@oleitao Actually, I found one last issue 😆. The library items should be added to the library section every time the user launches the app, they should get saved to the pinned items file.

**Comment by oleitao** [comment]
> @oleitao Actually, I found one last issue 😆. The library items should be added to the library section every time the user launches the app, they shouldn't get saved to the pinned items file. It might be a good idea to move this code out of SidebarPinnedModel.

First time app startup are loaded into library section (Desktop, Documents and Downloads), then each time user explore file system and open an file his path is add to most recent used folders (StorageApplicationPermissions.MostRecentlyUsedList) with limit of 3 folders. So, as example when "folder1" is loaded into most recent folders (Favorites) will be removed "Desktop" folder of startup default folders (Desktop, Documents and Downloads). When "folder2" is loaded into most recent folders will be removed "Documents" and replaced by one of the most recent used folders.

I don't understand what you are suggesting on Library section? On previous conversation you suggest me to do:
> @oleitao For now, I think documents, downloads, music, pictures and videos should be enough. Later on, we can detect all libraries.

**Comment by yaichenbaum** [comment]
@oleitao I am a little confused how this PR works, all I thought we were doing was adding an optional section to the sidebar for library items. Can you provide details on what's actually happening?

**Comment by oleitao** [comment]
> @oleitao I am a little confused how this PR works, all I thought we were doing was adding an optional section to the sidebar for library items. Can you provide details on what's actually happening?


**Favorites section**
Sorry, i make a mistake on my last comment please replace "library" word by "favorites"

> First time app startup are loaded into **library** section (Desktop, Documents and Downloads), then each time user explore file system and open an file his path is add to most recent used folders (StorageApplicationPermissions.MostRecentlyUsedList) with limit of 3 folders. So, as example when "folder1" is loaded into most recent folders (Favorites) will be removed "Desktop" folder of startup default folders (Desktop, Documents and Downloads). When "folder2" is loaded into most recent folders will be removed "Documents" and replaced by one of the most recent used folders.



**Library section**
On library section i is loading the follow default folders:

- DesktopPath
- DownloadsPath
- DocumentsPath
- PicturesPath
- MusicPath
- VideosPath

**Comment by yaichenbaum** [comment]
@oleitao Are the items in the library section stored in an ini file or are they added no matter what every time the app is opened?

**Comment by oleitao** [comment]
> @oleitao Are the items in the library section stored in an ini file or are they added no matter what every time the app is opened?
> 
> > then each time user explore file system and open an file his path is add to most recent used folders (StorageApplicationPermissions.MostRecentlyUsedList) with limit of 3 folders. So, as example when "folder1" is loaded into most recent folders (Favorites) will be removed "Desktop" folder of startup default folders (Desktop, Documents and Downloads). When "folder2" is loaded into most recent folders will be removed "Documents" and replaced by one of the most recent used folders.
> 
> @oleitao I don't think this is the behavior we decided to use when we originally discussed how the favorites section would work.


I suggest you to test this feature and will check that is everything ok. 

**Favorite section**
On that section if doesn’t exist any most recent folders will loaded the follow items:

- Desktop
- Downloads
- Documents
Otherwise, if exist most recent folders (last most recent 3 folders) their will be loaded into favourite section. 

**Library section**
> Are the items in the library section stored in an ini file or are they added no matter what every time the app is opened? 

That section is simple all content is loaded of “PinnedItems.json” into that section:

- Desktop
- Downloads
- Documents
- Pictures
- Music
- Videos

**Comment by yaichenbaum** [comment]
@oleitao Can you undo the changes to the favorites section? We don't want it to show recent folders, it's meant to be static. Regarding the library section, instead of loading the items from `PinnedItems.json`, can you add the items from code behind every time the user launches Files?

**Comment by oleitao** [comment]
> @oleitao Can you undo the changes to the favorites section? 
yes

> Regarding the library section, instead of loading the items from `PinnedItems.json`, can you add the items from code behind every time the user launches Files?

What do you mean by code behind, can you explain me?

**Comment by oleitao** [comment]
> > @oleitao Can you undo the changes to the favorites section?
> > yes
> 
> > Regarding the library section, instead of loading the items from `PinnedItems.json`, can you add the items from code behind every time the user launches Files?
> 
> What do you mean by code behind, can you explain me?

**Comment by yaichenbaum** [comment]
@oleitao By code behind, I mean to manually add the items to the section without loading them from the json file.

**Comment by yaichenbaum** [comment]
@oleitao I can work on the tweaking the library section, do you mind I push a few commits to your PR?

**Comment by oleitao** [comment]
> @oleitao By code behind, I mean to manually add the items to the section without loading them from the json file.

yes, i can pin item into library section manually without load them from .json file.
However, if i don't load from json on startup that section will be empty without items. 
Whats is that behaviour on startup if i don't load of .json file?

**Comment by oleitao** [comment]
> @oleitao I can work on the tweaking the library section, do you mind I push a few commits to your PR?

Yes, no problem

**Comment by yaichenbaum** [comment]
> Whats is that behaviour on startup if i don't load of .json file?

We will still add the items, just from the code instead of from the json file.

**Comment by yaichenbaum** [comment]
@oleitao The basic logic is working correctly now, do you want to work on these last issues in a separate PR?
- The app requires a restart after hiding or showing the library section
- Right clicking on a library item shows the option to unpin them and that's not the correct behavior.

**Comment by tsvietOK** [comment]
Is this solution back compatible with the previous one? I mean is it imports items from PinnedItems.json?

**Comment by yaichenbaum** [comment]
@tsvietOK This shouldn't affect items that were already pinned. The only change is the items that are pinned by default.

**Comment by gave92** [comment]
There are a few issues with this..
1. Pinned folders are not saved
2. You can add the same folder to the sidebar multiple times (but when you restart all pinned folders are lost, see [1])
3. If you delete the PinnedItems.json file on the next startup libraries folders will be added to the first sidebar section (not to the library section)

**Comment by yaichenbaum** [comment]
@gave92 Thanks for catching those, can you add them to #3869?
